### ISSUE TITLE
Attention mask computation during training

### ISSUE THREAD

**Issue by mtchiu2** [raoyongming/DynamicViT#11]
Hello,

Thank you for your work. I'm reading your implementation code for computing the attention mask. In lvvit.py line 665, you used F.gumbel_softmax(pred_score) to find the binary mask, but the last layer for computing pred_score is a log_softmax (line 529, same file). Shouldn't it be just a linear unit rather than a log_softmax, since F.gumbel_softmax takes logit rather than probability?

Thank you.

**Comment by raoyongming** [comment]
Hi @mtchiu2,

Thanks for your interest in our work. According to the Pytorch document, the input of `F.gumbel_softmax` should be [unnormalized log probabilities](https://pytorch.org/docs/stable/generated/torch.nn.functional.gumbel_softmax.html?highlight=gumbel#torch.nn.functional.gumbel_softmax). By comparing [the original paper of Gumbel Softmax](https://arxiv.org/pdf/1611.01144.pdf) and [the implementation in Pytorch](https://github.com/pytorch/pytorch/blob/b27dfa70c40f01cabaeea4063c382de8f624b327/torch/nn/functional.py#L1844), the `logits` should be the `log pi` in Eq.1 and Eq.2 of the Gumbel Softmax paper, which is the log of class probabilities. Therefore, we use `log_softmax` to compute the `logits`.

**Comment by mtchiu2** [comment]
I see. Thank you very much for the quick response.
